### PR TITLE
[dynamo] Put cells into `closure_cells` and document relevant parts

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140155
* #140154
* #140153
* #140435
* #140436
* #140152
* #140330
* #140202
* #140201
* #140151
* #140150
* #140149
* #140036
* __->__ #140035

This patch establishes the invariant that `ClosureVariable` and
`NewCellVariable` are always in `closure_cells`, never in
`symbolic_locals`, and therefore removes some duplicated code paths.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/functions.py [hunk 1]
```diff
@@ -289,7 +306,7 @@ def bind_args(self, parent, args, kwargs):
                         contents_var,
                     )
 
-                result[name] = out
+                closure_cells[name] = out
```
Comment (StrongerXi) [inline torch/_dynamo/variables/functions.py:309]: This is the critical change, everything else is documentation or removing the now dead code paths.

**Comment by StrongerXi** [comment]
Rebase

**Review by jansel — APPROVED** [review]

**Comment by StrongerXi** [comment]
Rebase.

**Comment by StrongerXi** [comment]
Rebase

**Pull request closed without merging**